From dc0ffec103961fd091ee1fc01a59859728ab216f Mon Sep 17 00:00:00 2001
From: Axel Naumann <Axel.Naumann@cern.ch>
Date: Thu, 16 Apr 2020 12:23:25 +0200
Subject: [PATCH 57/59] Preserve AnonStructIds of mangling ctx:

Across transactions, lambda types ($_0 etc) must keep their
mangling, instead of resetting the counter and recreating the
AnonStructIds. Enable that by swapping old and new builder
cxxabi mangle context. Fixes ROOT-10689.
---
 lib/CodeGen/ModuleBuilder.cpp | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/clang/lib/CodeGen/ModuleBuilder.cpp b/clang/lib/CodeGen/ModuleBuilder.cpp
index 72a0a134ca..748ed13fff 100644
--- a/clang/lib/CodeGen/ModuleBuilder.cpp
+++ b/clang/lib/CodeGen/ModuleBuilder.cpp
@@ -12,6 +12,7 @@
 
 #include "clang/CodeGen/ModuleBuilder.h"
 #include "CGDebugInfo.h"
+#include "CGCXXABI.h"
 #include "CodeGenModule.h"
 #include "clang/AST/ASTContext.h"
 #include "clang/AST/DeclObjC.h"
@@ -28,6 +29,14 @@
 using namespace clang;
 using namespace CodeGen;
 
+namespace {
+  struct CXXABICtxSwapper: clang::CodeGen::CGCXXABI {
+    void SwapCtx(clang::CodeGen::CGCXXABI &other) {
+      std::swap(MangleCtx, ((CXXABICtxSwapper&)other).MangleCtx);
+    }
+  };
+}
+
 namespace clang {
   class CodeGeneratorImpl : public CodeGenerator {
     DiagnosticsEngine &Diags;
@@ -157,6 +166,8 @@ namespace clang {
              && "Newly created module should not have weakRefRefs");
       Builder->WeakRefReferences.swap(OldBuilder->WeakRefReferences);
 
+      ((CXXABICtxSwapper&)*Builder->ABI).SwapCtx(*OldBuilder->ABI);
+
       return M.get();
     }
 
-- 
2.34.1

